execute/aggregate.go

```diff
@@ -102,6 +104,29 @@ func (t *aggregateTransformation) Process(id DatasetID, tbl flux.Table) error {
 
 func (t *aggregateTransformation) processChunk(chunk table.Chunk) error {
 	state, _ := t.d.Lookup(chunk.Key())
+	// XXX: sean (Oct 7, 2021) - I am not entirely sure why, but checking `ok`
+	// here before setting `state` to the return value of `popFreeState()` avoids
+	// some major bugs. Without it, some aggregates produce wrong answers, and tests
+	// start to fail.
+	//
+	// My hunch is that while `multiState` and `aggregateState` implement
+	// `Recyclable`, the underlying aggregate may not. So we append states to
+	// `t.freeStates` (see the `flushKey` function) in all circumstances, even
+	// if the underlying aggregate cannot be recycled. This causes some aggregates
+	// (sum in particular) to re-use state erroneously.
+	//
+	// Still, I don't know why checking if it's recyclable here solves the problem.
+	// If the state wasn't recyclable, t.freeStates should be empty anyway, so this check
+	// shouldn't make a difference.
+	//
+	// Also, if `state` is nil, wouldn't `ok` always be false? In which case, the below
+	// `if` block would never trigger, which means that the fix would not work.
+	// Something strange is going on here.
+	_, ok := state.(Recyclable)
+	if ok && state == nil {
+		// If no reusable state is available, state will remain nil
+		state = t.popFreeState()
+	}
```

Comment:
Here is how I understand whats going on.

First, in the flushKey function below we only adds states to the freeStates list of its recyclable. 
Second, as your comment clarifies the t.popFreeState() function will return `nil` if there are no free states.

My hunch here is that we are overriding a known good state with `nil` and thus getting wrong answers.

This is how I would expect this code to work:

```go
state, ok := t.d.Lookup(chunk.Key())
if !ok {
    // We did not find a previous state, therefore we should try and resuse a cached one if it exists.
    // If no reusable state is available, state will remain nil.
    // If state remains nil, a new state will be created below.
    state = t.popFreeState()
}
// the rest of the function, no need to check if the state is recyclable here.
```